crates/federate/src/send.rs

```diff
@@ -0,0 +1,147 @@
+/// 3. The main fail_count should only be updated under certain conditions, which are best handled
+///    in the main task.
+/// 4. It maintains consistency in how data is communicated (all via channels rather than a mix of
+///    channels and atomics).
+/// 5. It simplifies concurrency management and makes the flow of data more predictable.
+pub(crate) enum SendActivityResult {
+  Success(SendSuccessInfo),
+  Failure { fail_count: i32 },
+}
+/// Represents a task for retrying to send an activity.
+///
+/// This struct encapsulates all the necessary information and resources for attempting
+/// to send an activity to multiple inbox URLs, with built-in retry logic.
+pub(crate) struct SendRetryTask<'a> {
+  pub activity: &'a SentActivity,
+  pub object: &'a SharedInboxActivities,
+  /// Must not be empty at this point
+  pub inbox_urls: Vec<Url>,
+  /// Channel to report results back to the main instance worker
+  pub report: &'a mut UnboundedSender<SendActivityResult>,
+  /// The first request will be sent immediately, but subsequent requests will be delayed
+  /// according to the number of previous fails + 1
+  ///
+  /// This is a read-only immutable variable that is passed only one way, from the main
+  /// thread to each send task. It allows the task to determine how long to sleep initially
+  /// if the request fails.
+  pub initial_fail_count: i32,
+  /// For logging purposes
+  pub domain: String,
+  pub context: Data<LemmyContext>,
+  pub stop: CancellationToken,
+}
+
+impl<'a> SendRetryTask<'a> {
+  // this function will return successfully when (a) send succeeded or (b) worker cancelled
+  // and will return an error if an internal error occurred (send errors cause an infinite loop)
+  pub async fn send_retry_loop(self) -> Result<()> {
+    let SendRetryTask {
+      activity,
+      object,
+      inbox_urls,
+      report,
+      initial_fail_count,
+      domain,
+      context,
+      stop,
+    } = self;
+    debug_assert!(!inbox_urls.is_empty());
+
+    let pool = &mut context.pool();
+    let Some(actor_apub_id) = &activity.actor_apub_id else {
+      return Err(anyhow::anyhow!("activity is from before lemmy 0.19"));
+    };
+    let actor = get_actor_cached(pool, activity.actor_type, actor_apub_id)
+      .await
+      .context("failed getting actor instance (was it marked deleted / removed?)")?;
+
+    let object = WithContext::new(object.clone(), FEDERATION_CONTEXT.deref().clone());
+    let requests = SendActivityTask::prepare(&object, actor.as_ref(), inbox_urls, &context).await?;
+    for task in requests {
+      // usually only one due to shared inbox
+      tracing::debug!("sending out {}", task);
+      let mut fail_count = initial_fail_count;
+      while let Err(e) = task.sign_and_send(&context).await {
+        fail_count += 1;
+        report.send(SendActivityResult::Failure {
+          fail_count,
+          // activity_id: activity.id,
+        })?;
```

Comment:
I would try to send this only once per activity, otherwise its awkward to handle.